lib/ui/highlight-manager.js

```diff
@@ -9,36 +9,143 @@ function HighlightManager(opts) {
+    // rely on setTimeout to steal the highlight: if 30 row rendering calls are
+    // in the event loop ahead of this code, the blue highlight will very
+    // noticeably flicker on the top result row. To avoid this, we steal the
+    // highlight immediately, then use rAF to steal it just before the next
+    // paint, then set *another* rAF inside the first rAF, preventing another
+    // paint from showing a result in the list as highlighted. This seems a
+    // little excessive, but actually the inner rAF seems to be the key to
+    // truly minimizing the flickering as we steal the highlight.
+    this.recommendation.isHighlighted = true;
+    this.popup.selectedIndex = -1;
+    this.win.requestAnimationFrame(() => {
+      this.recommendation.isHighlighted = true;
+      this.popup.selectedIndex = -1;
+      this.win.requestAnimationFrame(() => {
+        this.recommendation.isHighlighted = true;
+        this.popup.selectedIndex = -1;
+      });
+    });
+  },
+  clearHighlight: function() {
+    const resultsContainer = this.win.document.getAnonymousElementByAttribute(this.popup, 'anonid', 'richlistbox');
+    const resultRows = resultsContainer.getElementsByClassName('autocomplete-richlistitem');
+    Array.prototype.forEach.call(resultRows, row => { row.selected = false; });
+  },
+  // Due to constraints in combining XBL and JS, we have to totally take over
+  // highlight management for the list of results as well as the recommendation.
   adjustHighlight: function(evt) {
-    // if the recommendation isn't shown, do nothing: let the popup manage focus
-    // otherwise,
-    // check if the recommendation is shown + highlighted
-    // check the selectedIndex of the popup
-    // if it's a down key, move the focus downward, wrapping around if needed
-    // if it's an up key, move the focus upward, wrapping around if needed
+    // Batch all DOM access here at the start of the function.
+    const resultsContainer = this.win.document.getAnonymousElementByAttribute(this.popup, 'anonid', 'richlistbox');
+    // TODO: should this be var, not const?
+    const resultRows = resultsContainer.getElementsByClassName('autocomplete-richlistitem');
+    // resultRows is a live collection, and the XBL code inserts elements over
+    // several turns, so we'll use listLength for calculating the past state
+    // of the world, but when we want to assign focus to the last item in the
+    // list, we'll use the live collection.
+    const listLength = resultRows.length;
+    const selectedIndex = this.popup.selectedIndex;
+    const recommendationVisible = this.recommendation.el && !this.recommendation.el.collapsed;
+    const recommendationHighlighted = this.recommendation.el && this.recommendation.isHighlighted;
+
+    // Clear all highlights. The DOM is now dirty and not trustworthy.
+    this.clearHighlight();
+
+    // Start by checking what's currently highlighted, and alter the state:
+
+    // recommendation
```

Comment:
Nit: can we refactor comments to start with a capital letter and be a complete sentence?